File: holec_trading/holec_trading/page/buy_ticket/buy_ticket.js
frappe.pages['buy_ticket'].on_page_load = function(wrapper) {
    let page = frappe.ui.make_app_page({ parent: wrapper, single_column: true });
    $(wrapper).find('.page-head').hide();
    $(frappe.render_template('buy_ticket', {})).appendTo(page.main);

    // Fetch dropdown values & initial date
    frappe.call({
        method: 'holec_trading.holec_trading.page.buy_ticket.buy_ticket.get_ticket_init_data',
        callback: function(r) {
            if (!r.message) return;
            const { suppliers, unapproved_count, items, today_date } = r.message;

            $('#supplier-unapproved-count').text(unapproved_count);

            let supplierOptions = suppliers.map(s => `
                <option value="${s.name}">${s.supplier_name || s.name} (${s.supplier_group || 'General'})</option>
            `);
            $('#sel-supplier').append(supplierOptions.join(''));

            if (items && items.length) {
                let itemOptions = items.map(i => `<option value="${i.name}">${i.item_name || i.name}</option>`);
                $('#sel-item').html(itemOptions.join(''));
            }

            let todayFormatted = frappe.datetime.str_to_user(today_date || frappe.datetime.get_today());
            $('#inp-delivery-date').val(todayFormatted);
        }
    });

    // Handle form submit
    $('#create-buy-ticket-form').on('submit', function(e) {
        e.preventDefault();

        let supplier = $('#sel-supplier').val();
        let item = $('#sel-item').val();
        let qty = $('#inp-quantity').val();
        let dateVal = $('#inp-delivery-date').val();

        if (!qty) {
            frappe.show_alert({ message: __('Enter an expected quantity'), indicator: 'orange' });
            return;
        }

        frappe.call({
            method: 'holec_trading.holec_trading.page.buy_ticket.buy_ticket.submit_new_ticket',
            args: {
                supplier: supplier,
                item: item,
                expected_qty: qty,
                delivery_date: frappe.datetime.user_to_str(dateVal)
            },
            freeze: true,
            callback: function(r) {
                if (r.message) {
                    frappe.show_alert({ message: __('Ticket created successfully'), indicator: 'green' });
                    // Navigate back to the Lots page
                    frappe.set_route('lot_');
                }
            }
        });
    });

    // Cancel Button
    $('#btn-cancel-ticket').on('click', function() {
        frappe.set_route('lot_');
    });
};

File: holec_trading/holec_trading/page/buy_ticket/buy_ticket.py
import frappe
from frappe.utils import nowdate, random_string

@frappe.whitelist()
def get_ticket_init_data():
    # Only show Approved suppliers
    approved_suppliers = frappe.get_all(
        "Supplier",
        filters={"custom_workflow_status": "Approved", "disabled": 0},
        fields=["name", "supplier_name", "supplier_group"]
    )

    unapproved_count = frappe.db.count("Supplier", {
        "custom_workflow_status": ["!=", "Approved"]
    })

    items = frappe.get_all("Item", filters={"is_purchase_item": 1}, fields=["name", "item_name"])
    if not items:
        items = [{"name": "Maize", "item_name": "Maize"}, {"name": "Beans", "item_name": "Beans"}]

    return {
        "suppliers": approved_suppliers,
        "unapproved_count": unapproved_count,
        "items": items,
        "today_date": nowdate()
    }

@frappe.whitelist()
def submit_new_ticket(supplier, item, expected_qty, delivery_date):
    if not supplier:
        frappe.throw("Please select a Supplier.")
    if not expected_qty:
        frappe.throw("Please enter expected quantity.")

    # 1. Create Buy Ticket Doc
    ticket = frappe.new_doc("Buy Ticket")
    if frappe.db.has_column("Buy Ticket", "supplier"):
        ticket.supplier = supplier
    if frappe.db.has_column("Buy Ticket", "item"):
        ticket.item = item
    if frappe.db.has_column("Buy Ticket", "expected_quantity"):
        ticket.expected_quantity = expected_qty
    if frappe.db.has_column("Buy Ticket", "expected_delivery_date"):
        ticket.expected_delivery_date = delivery_date
    ticket.insert(ignore_permissions=True)

    # 2. Create Lot Document linked to this ticket
    lot = frappe.new_doc("Lot")
    lot_code = "LOT-" + random_string(5).upper()
    lot.name = lot_code
    
    if frappe.db.has_column("Lot", "ticket"):
        lot.ticket = ticket.name
    if frappe.db.has_column("Lot", "supplier"):
        lot.supplier = supplier
    if frappe.db.has_column("Lot", "quantity"):
        lot.quantity = expected_qty
    if frappe.db.has_column("Lot", "state"):
        lot.state = "Ticket"
    lot.insert(ignore_permissions=True)

    frappe.db.commit()
    return {"ticket": ticket.name, "lot": lot.name}